docs/rfc-1-staking-contract.adoc

```diff
@@ -0,0 +1,381 @@
+indeed at stake.
+
+Beneficiary:: An address where the rewards for participation are sent, earned by
+the operator, on behalf of the owner. A beneficiary doesn’t sign or publish any
+protocol-relevant transactions, but any currency or tokens earned by the
+operator will be transferred to the beneficiary.
+
+Authorizer:: An address appointed by an owner to authorize applications on
+behalf of them. An application must be approved by the authorizer before the
+operator is eligible to participate.
+
+Delegated stake:: An owner’s staked tokens, delegated to the operator by the
+owner. Delegation enables KEEP owners to have their wallets offline and their
+stake operated by operators on their behalf.
+
+== Specification
+
+=== Functionality
+
+==== Delegating a stake
+
+Tokens are delegated by the owner. During the delegation, the owner needs to
+appoint an operator, beneficiary, and authorizer. The owner may decide to
+delegate just a portion of their tokens. The owner may delegate multiple times
+to different operators. For simplicity, the operator address can not be reused
+between delegations even if the delegation has been recovered. The only
+exception is when the delegation has been canceled. To protect against griefing
+by frontrunning stake delegation transactions and reserving operator addresses,
+staking contract should require a governable minimum stake amount for each
+delegation.
+
+Staking contract binds the owner with the operator, beneficiary, and authorizer
+for the given stake delegation and records the stake delegation amount.
+Delegation is set to active after 2 hours, once the cancellation window passes.
+
+Existing KEEP and NU stake owners also need to perform a delegation so that T
+staking contract can record their operator, beneficiary, and authorizer. Since
+these values are read-only, they should be copied from KEEP staking contract to
+T staking contract and NU stakers should set them during the delegation. For
+existing stakes, the entire stake amount is registered in T staking contract.
+
+There are 2 hours to cancel a stake without suffering an undelegation period.
+This functionality is needed e.g. in case of a typo in the operator address.
+Undelegation can not be canceled in the same transaction in which it was
+created. Once the delegation has been canceled, the operator address can be used
+again.
+
+==== Authorizing an application
+
+Before the operator is eligible to participate in the given application, the
+authorizer appointed during the stake delegation needs to review the application
+and approve it to use the stake. From the moment of approval, the application
+is authorized to slash or seize the stake.
+
+The authorizer decides up to which amount the given application is eligible to
+slash. This allows managing the risk profile of the staker individually for each
+application. The authorized stake amount can only be increased.
+
+Authorization can not be canceled.
```

Comment:
This part needs clarification because it should be possible to deallocate, following pertinent application- and stake-level restrictions (e.g., satisfying specific deallocation periods).

Reply:
Allowing to cancel the authorization makes the process a little more complex. Canceling authorization per application means that every application should have it is own rules about the "cooldown" period. 

For example, for the random beacon, you need to initialize deauthorization and your stake is no longer eligible for new work selection. After 2 months of deauthorization period, random beacon can no longer slash the stake. 
For tBTC v2, the deauthorization period could be one month and for PRE one week (just an example).
In this model, the stake would get unlocked as soon as the last deauthorization process completes and we wouldn't need a global undelegation period. Deauthorization would happen on the staking contract level and the application would be notified about it. Also, probably we would not need the stake delegation cancellation timeout.

The model presented currently in this RFC is quite similar except that there is a global undelegation time that all applications need to be satisfied with and there is an assumption that authorized application needs to be trusted enough so that when operator stops using it without undelegating, it is not going to slash the stake.

Another - more sophisticated - option would be to let each authorized application lock a stake. Instead of having a hardcoded deauthorization period per application, the application would get notified about deauthorization request and it would trigger an internal cooldown process. Once that process completes - and whatever it means depends on the application - the application would unlock the stake on the staking contract. The stake would get  unlocked as soon as the last lock on the stake is cleared. This option is tempting but it's also way more complex given that each application can receive authorization for a different amount of tokens as [explained here](https://github.com/threshold-network/solidity-contracts/pull/15/files/c1e8a0ba7479ffb23fd6881d4bc08f113a8a2c2c#r703609027) by @eth-r.

Reply:
When each app has own cooldown period increases complexity only with iteration over apps to check time and value instead of one value (there is mvp in #7). And we will have panic button so trusting is not so bad at all. From our side we want partial deallocation/deauthorization so anyway it will increase complexity of this process and adding one parameter with time variable is not so big (in my opinion)

Reply:
(changing this mechanism is discussed in [this](https://github.com/threshold-network/solidity-contracts/pull/15#discussion_r703412443) thread)